runtime/compiler/x/amd64/codegen/AMD64JNILinkage.cpp

```diff
@@ -1181,14 +1181,25 @@ void TR::AMD64JNILinkage::cleanupReturnValue(
       // type so that we sign and zero extend the narrower integer return types properly.
       //
       TR_X86OpCodes op;
+      TR::SymbolReference      *callSymRef = callNode->getSymbolReference();
       TR::ResolvedMethodSymbol *callSymbol = callNode->getSymbol()->castToResolvedMethodSymbol();
       TR_ResolvedMethod *resolvedMethod = callSymbol->getResolvedMethod();
 
       bool isUnsigned = resolvedMethod->returnTypeIsUnsigned();
+
       switch (resolvedMethod->returnType())
          {
          case TR::Int8:
-            if (isUnsigned)
+            if (comp()->getSymRefTab()->isReturnTypeBool(callSymRef))
+               {
+               // For bool return type, must check whether value returned by
+               // JNI is zero (false) or non-zero (true) to yield Java result
+               generateRegRegInstruction(TEST1RegReg, callNode,
+                     linkageReturnReg, linkageReturnReg, cg());
+               generateRegInstruction(SETNE1Reg, callNode, linkageReturnReg, cg());
+               op = TR::Compiler->target.is64Bit() ? MOVZXReg8Reg1 : MOVZXReg4Reg1;
```

Comment:
I like the way you keep this consistent with the rest of codes.
Just a side note: I'm not sure if checking `TR::Compiler->target.is64Bit()` as this file is only consumed by 64-bit JIT. We may want to revisit this file later and cleanup the unnecessary bitness check.

Reply:
Yes, I was surprised by the number of places in this file that check `target.is64Bit()`.  Were developers thinking at some point earlier on that there could be just one version of the code gen for x86 for both 64 bit and 32 bit?  That's how things seem to be arranged for Power.

Should I open an issue to track that suggestion for cleanup?

Reply:
Yup, I think an Issue to track the cleanup is good for now.

Reply:
Created issue #4140